src/state.rs

```diff
@@ -341,9 +388,49 @@ impl DemandProxyState {
                     .on_demand_dns_cache_misses
                     .get_or_create(&labels)
                     .inc();
-                // TODO: optimize so that if multiple requests to the same hostname come in at the same time,
+                // optimize so that if multiple requests to the same hostname come in at the same time,
                 // we don't start more than one background on-demand DNS task
-                Self::resolve_on_demand_dns(self.to_owned(), workload).await;
+                if state
+                    .get_cached_resolve_dns_for_hostname(&hostname)
+                    .is_none()
+                {
+                    let cached_resolve_dns = SingleFlight::new();
+                    state.set_cached_resolve_dns_for_hostname(
+                        hostname.to_owned(),
+                        cached_resolve_dns,
+                    );
+
+                    let element = state.get_cached_resolve_dns_for_hostname(&hostname);
+                    match element {
+                        Some(sf_map_element) => {
+                            // Check if this is the first request of doing DNS task for hostname
+                            if sf_map_element.calling_index == 1 {
+                                // doing the dns tasks
+                                Self::resolve_on_demand_dns(self.to_owned(), workload).await;
+                                // need to remove the element in sf_by_hostname because the DNS resolving task
+                                // is completed and the IP addresses already are stored in Hashmap of by_hostname
+                                state.remove_cached_resolve_dns_for_hostname(&hostname);
+                                sf_map_element.notify_waiters();
+                            } else {
+                                sf_map_element.wait_for_notifying();
```

Comment:
@zhlsunshine need to confirm whether it can be waken up after a previous notify, otherwise there is a race this thread can be blocking forever

Reply:
Hi @hzxuzhonghu, yes, I confirm that it can be waken up after a previous notify. This is the common usage of A Condition Variable recorded in official sites of Rust [here](https://doc.rust-lang.org/stable/std/sync/struct.Condvar.html#:~:text=Condition%20variables%20are%20typically%20associated,the%20current%20thread%20of%20execution.)

Reply:
I think quite the opposite, the example actually does notify after wait

```
#![allow(unused)]
fn main() {
    use std::sync::{Arc, Condvar, Mutex};
    use std::{thread, time};

    let pair = Arc::new((Mutex::new(false), Condvar::new()));
    let pair2 = Arc::clone(&pair);

    // Inside of our lock, spawn a new thread, and then wait for it to start.
    thread::spawn(move || {
        println!("in thread");

        let (lock, cvar) = &*pair2;
        let mut started = lock.lock().unwrap();
        *started = true;
        // We notify the condvar that the value has changed.
        println!("notify");
        cvar.notify_one();
    });

    // Wait for the thread to start up.
    thread::yield_now();
    let ten_millis = time::Duration::from_millis(10);
    thread::sleep(ten_millis);
    let (lock, cvar) = &*pair;
    let mut started = lock.lock().unwrap();
    println!("waiting");
    started = cvar.wait(started).unwrap();
    println!("waiting return");
}

```

Trey this example, you can see it never end

Reply:
Hi @hzxuzhonghu, I have tried above code in my environment, so you care a case which make sure the notifying operation occur before the waiting operation. Yeah, it would be suspended in main thread.  For addressing this, I list 2 solution about it:

1. In our case, the DNS Resolving tasks can not be ahead of the waiting requests, so this can reduce the chance of suspending
2. Using `wait_timeout` instead of `wait`, and the timeout can be 100 milliseconds

Is this okay?

Reply:
Hi @hzxuzhonghu, I have tried above code in my environment, so you create a case which make sure the notifying operation occur before the waiting operation. Yeah, it would be suspended in main thread.  For addressing this, I list 2 solution about it:

1. In our case, the DNS Resolving tasks can not be ahead of the waiting requests, so this can reduce the chance of suspending
2. Using `wait_timeout` instead of `wait`, and the timeout can be 100 milliseconds

Is this okay?

Reply:
Hi @hzxuzhonghu, I have tried above code in my environment, yeah, it would be suspended in main thread. So you create a case which make the notifying operation occurred before the waiting operation. For addressing this, I list 2 solution about it:
1. In our case, the DNS Resolving tasks can not be ahead of the waiting requests, so this can reduce the chance of suspending
2. Using `wait_timeout` instead of `wait`, and the timeout can be 100 milliseconds

Is this okay?

Reply:
Hi @hzxuzhonghu, I use above code you provided and change them based on the solution, the final code is below:

```
#![allow(unused)]
fn main() {
    use std::sync::{Arc, Condvar, Mutex};
    use std::{thread, time};

    let pair = Arc::new((Mutex::new(false), Condvar::new()));
    let pair2 = Arc::clone(&pair);

    // Inside of our lock, spawn a new thread, and then wait for it to start.
    thread::spawn(move || {
        println!("in thread");

        let (lock, cvar) = &*pair2;
        let mut started = lock.lock().unwrap();
        *started = true;
        // We notify the condvar that the value has changed.
        println!("notify");
        cvar.notify_all();
    });

    // Wait for the thread to start up.
    thread::yield_now();
    let ten_millis = time::Duration::from_millis(100);
    thread::sleep(ten_millis);
    let (lock, cvar) = &*pair;
    let mut started = lock.lock().unwrap();
    println!("waiting");
    // Use wait_timeout with a timeout of 100 milliseconds
    let (new_started, _timeout_result) = cvar.wait_timeout(started, ten_millis).unwrap();
    // Check the condition after waiting
    if *new_started {
        println!("Thread started.");
    } else {
        println!("Timeout: Thread did not start.");
    }
    println!("waiting return");
}
```

The output looks like below:
![image](https://github.com/istio/ztunnel/assets/4101246/c6a7db04-24ea-4e8a-a2dd-68631e267b09)

Reply:
I am not sure this is a good way, and also whether 100ms is the right timeout

Reply:
Hi @hzxuzhonghu, basically, the DNS resolving task can last from microseconds to milliseconds which is depended on many facts, and I think 100 milliseconds should be a conservative value for this.